### PR TITLE
resolve #21056 - Add support for full cloning a Qemu VM. Fix some issues. Update doc

### LINKED ISSUE #21056: proxmox_kvm : delete args
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

proxmox_kvm

##### ANSIBLE VERSION

```
2.2.1.0
```

plus

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/misc/proxmox_kvm.py

##### SUMMARY

Deleting ```args``` does not work.

##### STEPS TO REPRODUCE

```
  tasks:

  - name: Remove
    proxmox_kvm:
      api_user: "{{ api_user }}"
      api_password: "{{ api_password }}"
      api_host: "{{ api_host }}"
      node: "{{ node }}"
      name: "{{ vmname }}"
      delete: args
```

##### EXPECTED RESULTS

```args``` get's removed from vm config.

##### ACTUAL RESULTS

Playbook exits successful, but the args are stil in the vm config.

### PR DESCRIPTION
**Pull request by helldorado**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

modules/cloud/misc/proxmox_kvm.py

##### ANSIBLE VERSION

```yaml
ansible --version
ansible 2.2.0.0
  config file = /usr/local/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

Fixes #21056
Add new feature
  - Clone

```yaml
---
- hosts: orchestrator
  remote_user: root
  gather_facts: False
  vars:
    instance:
      - name: spynal
        proxmox:
          api_host  : labhyp.helldorado.info
          node      : labhyp1
          cores     : 2
          vcpus     : 1
          net       : '{"net0":"virtio,bridge=vmbr0"}'
          virtio    : '{"virtio0":"VMs:50,format=qcow2"}'
```

```yaml
- name: CLONE
  local_action:
    module: proxmox_kvm
    api_user     : "{{ item.proxmox.api_user|default('root@pam') }}"
    api_password : "{{ item.proxmox.api_password|default(omit) }}"
    api_host     : "{{ item.proxmox.api_host }}"
    node         : "{{ item.proxmox.node }}"
    name         :  "{{ item.name }}"
    clone        : zavala
    storage      : 'VMs'
    timeout      : 300
  with_items: "{{ instance }}"
```

> ℹ️  Some output has truncated

```yaml
ansible@spynal-orchestrator:/usr/local/ansible/orchestrator (master)$ ansible-playbook playbooks/servers.yml -vvvv
PLAYBOOK: servers.yml **********************************************************
1 plays in playbooks/servers.yml

PLAY [orchestrator] ************************************************************

TASK [deploy : Include virtual] ************************************************
task path: /usr/local/ansible/orchestrator/roles/deploy/tasks/main.yml:4
included: /usr/local/ansible/orchestrator/roles/deploy/tasks/virtual.yml for spynal-orchestrator

TASK [deploy : CLONE] **********************************************************
task path: /usr/local/ansible/orchestrator/roles/deploy/tasks/virtual.yml:32
Using module file /usr/local/ansible/orchestrator/roles/deploy/library/proxmox_kvm.py
changed: [spynal-orchestrator -> localhost] => (item={{
        },
        "module_name": "proxmox_kvm"
    },
    "item": {
        "cores" : 2,
        "delete": "cpuunits",
        "name": "spynal",
        "net": "{\"net0\":\"virtio,bridge=vmbr0\"}",
        "proxmox": {
            "api_host": "labhyp.helldorado.info",
            "node": "labhyp1"
        },
        "vcpus": 1,
        "virtio": "{\"virtio0\":\"VMs:50,format=qcow2\"}"
    },
    "msg": "VM spynal with newid 136 cloned from vm with vmid 108"
}

PLAY RECAP *********************************************************************
spynal-orchestrator        : ok=2    changed=1    unreachable=0    failed=0
```

Issue #21056

###### Update
```yaml
---
- hosts: orchestrator
  remote_user: root
  gather_facts: False
  vars:
    instance:
      - name: spynal
        proxmox:
          api_host : labhyp.helldorado.info
          node     : labhyp1
          cores    : 4
          vcpus    : 1
          net      : '{"net0":"virtio,bridge=vmbr0"}'
          virtio   : '{"virtio0":"VMs:50,format=qcow2"}'
          update   : yes
```

```yaml
- name: CREATE/UPDATE VM
  local_action:
    module: proxmox_kvm
    api_user      : "{{ item.proxmox.api_user|default('root@pam') }}"
    api_password  : "{{ item.proxmox.api_password|default(omit) }}"
    api_host      : "{{ item.proxmox.api_host }}"
    node          : "{{ item.proxmox.node }}"
    name          : "{{ item.name }}"
    kvm           : "{{ item.kvm|default('yes') }}"
    net           : '{"net0":"virtio,bridge={{ vlan|default(pxe.vlan) }}"}'
    virtio        : "{{ item.virtio }}"
    cores         : "{{ item.cores|default('4') }}"
    sockets       : "{{ item.sockets|default('1') }}"
    vcpus         : "{{ item.vcpus|default(omit) }}"
    memory        : "{{ item.memory|default('2048') }}"
    balloon       : "{{ item.ballon|default(omit) }}"
    hotplug       : "{{ item.hotplug|default('memory,cpu,disk,network') }}"
    numa_enabled  : "{{ item.numa_enabled|default('yes') }}"
    keyboard      : "{{ item.keyboard|default('fr') }}"
    state         : "{{ item.state|default('present') }}"
    update        : "{{ item.update|default('yes') }}"
    delete        : "{{ item.delete|default(omit) }}"
    cpuunits      : "{{ item.cpuunits|default(omit) }}"
    revert        : "{{ item.revert|default(omit) }}"
  register: qemu
  with_items: "{{ instance }}"
```

###### Delete

> Param `delete: 'template'` added in instance dict per example

```yaml
ansible@spynal-orchestrator:/usr/local/ansible/orchestrator (master)$ ansible-playbook playbooks/servers.yml -vvvvvvvvv
PLAYBOOK: servers.yml **********************************************************
1 plays in playbooks/servers.yml

PLAY [orchestrator] ************************************************************

TASK [deploy : Include virtual] ************************************************
task path: /usr/local/ansible/orchestrator/roles/deploy/tasks/main.yml:4
included: /usr/local/ansible/orchestrator/roles/deploy/tasks/virtual.yml for spynal-orchestrator

TASK [deploy : CREATE/UPDATE VM] ******************************************************
task path: /usr/local/ansible/orchestrator/roles/deploy/tasks/virtual.yml:4
Using module file /usr/local/ansible/orchestrator/roles/deploy/library/proxmox_kvm.py
changed: [spynal-orchestrator -> localhost] => (item={}) => {
    "changed": true,
    "invocation": {
        "module_args": {
        },
        "module_name": "proxmox_kvm"
    },
    "item": {
        "cores": 2,
        "cpuunits": 50000,
        "delete": "template",
        "name": "combo",
        "net": "{\"net0\":\"virtio,bridge=vmbr0\"}",
        "proxmox": {
            "api_host": "labhyp.helldorado.info",
            "node": "labhyp1"
        },
        "tag": 1488514,
        "update": true,
        "vcpus": 1,
        "virtio": "{\"virtio0\":\"VMs:50,format=qcow2\"}"
    },
    "msg": "Settings has deleted on VM combo with vmid 108"
}

PLAY RECAP *********************************************************************
spynal-orchestrator        : ok=2    changed=1    unreachable=0    failed=0
```

###### Revert

> Param `revert: 'cpuunits'` added in instance dict per example

```yaml
ansible@spynal-abah-orchestrator:/usr/local/ansible/orchestrator (master)$ ansible-playbook playbooks/servers.yml -vvvvvvvvv

PLAYBOOK: servers.yml **********************************************************
1 plays in playbooks/servers.yml

PLAY [orchestrator] ************************************************************

TASK [deploy : Include virtual] ************************************************
task path: /usr/local/ansible/orchestrator/roles/deploy/tasks/main.yml:4
included: /usr/local/ansible/orchestrator/roles/deploy/tasks/virtual.yml for spynal-orchestrator

TASK [deploy : CREATE/UPDATE VM] ******************************************************
task path: /usr/local/ansible/orchestrator/roles/deploy/tasks/virtual.yml:4
Using module file /usr/local/ansible/orchestrator/roles/deploy/library/proxmox_kvm.py
changed: [spynal-orchestrator -> localhost] => (item={}) => {
    "changed": true,
    "invocation": {
        "module_args": {
            },
        },
        "module_name": "proxmox_kvm"
    },
    "item": {
        "cores": 2,
        "cpuunits": 50000,
        "name": "combo",
        "net": "{\"net0\":\"virtio,bridge=vmbr0\"}",
        "proxmox": {
            "api_host": "labhyp.helldorado.info",
            "node": "labhyp1"
        },
        "revert": "cpuunits",
        "tag": 1488514,
        "update": true,
        "vcpus": 1,
        "virtio": "{\"virtio0\":\"VMs:50,format=qcow2\"}"
    },
    "msg": "Settings has reverted on VM combo with vmid 108"
}

PLAY RECAP *********************************************************************
spynal-orchestrator        : ok=2    changed=1    unreachable=0    failed=0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/misc/proxmox_kvm.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 
 ANSIBLE_METADATA = {'status': ['preview'],
                     'supported_by': 'community',
-                    'version': '1.0'}
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox_kvm.py]: This is the version of the metadata, not your file. So no need to change that. 

It's on my to do list to document ANSIBLE_METADATA so I'll add a note about this.

File: lib/ansible/modules/cloud/misc/proxmox_kvm.py [hunk 2]
```diff
@@ -99,6 +99,12 @@
     required: false
     default: null
     type: string
+  clone:
+    description:
+      - Name of VM to be cloned. If C(vmid) is setted, C(clone) can take arbitrary value but required for intiating the clone.
+    required: false
+    default: null
+    type: string
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox_kvm.py]: No need to document `type:` in the DOCUMENTION block.

File: lib/ansible/modules/cloud/misc/proxmox_kvm.py [hunk 3]
```diff
@@ -893,28 +1073,70 @@ def main():
...
   # If vmid not set get the Next VM id from ProxmoxAPI
   # If vm name is set get the VM id from ProxmoxAPI
-  if module.params['vmid'] is not None:
-    vmid = module.params['vmid']
-  elif state == 'present':
-    vmid = get_nextvmid(proxmox)
-  elif module.params['name'] is not None:
-    vmid = get_vmid(proxmox, name)[0]
+
+  if module.params['delete'] is not None and module.params['revert'] is not None:
```
Comment (gundalow) [inline lib/ansible/modules/cloud/misc/proxmox_kvm.py]: Their are built in helpers for argspec  for the this, have a look in other modules for "exclusive"

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by helldorado** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]

**Comment by mattclay** [comment]
CI failure due to PEP 8 issues:

```
2017-02-13 18:55:47 ERROR: PEP 8: lib/ansible/modules/cloud/misc/proxmox_kvm.py:900:5: E121 continuation line under-indented for hanging indent (legacy)
2017-02-13 18:55:47 ERROR: PEP 8: lib/ansible/modules/cloud/misc/proxmox_kvm.py:901:7: E121 continuation line under-indented for hanging indent (legacy)
```

I recommend running the PEP 8 checks locally with `make pep8` to easily see what needs changes.

**Comment by helldorado** [comment]
ready_for_review

**Pull request merged**